File: backend/app/services/categorization/rules.py
from __future__ import annotations

import re
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Transaction, Payee, Account, AutoCategorizationRule


async def apply_rules(db: AsyncSession, household_id: str):
    """Apply auto-categorization rules to uncategorized transactions."""
    rules_result = await db.execute(
        select(AutoCategorizationRule)
        .where(
            AutoCategorizationRule.household_id == household_id,
            AutoCategorizationRule.enabled.is_(True),
        )
        .order_by(AutoCategorizationRule.priority.desc())
    )
    rules = rules_result.scalars().all()
    if not rules:
        return

    txn_result = await db.execute(
        select(Transaction)
        .join(Account, Transaction.account_id == Account.id)
        .where(Account.household_id == household_id)
        .where(Transaction.category_id.is_(None))
        .where(Transaction.parent_transaction_id.is_(None))
    )
    transactions = txn_result.scalars().all()

    payee_cache: dict[str, str] = {}

    for txn in transactions:
        for rule in rules:
            matched = False
            target_value = ""

            if rule.match_field == "payee" and txn.payee_id:
                if txn.payee_id not in payee_cache:
                    p_result = await db.execute(select(Payee.name).where(Payee.id == txn.payee_id))
                    payee_cache[txn.payee_id] = p_result.scalar_one_or_none() or ""
                target_value = payee_cache[txn.payee_id]
            elif rule.match_field == "notes":
                target_value = txn.notes or ""
            elif rule.match_field == "amount":
                target_value = str(txn.amount)

            if rule.match_type == "contains":
                matched = rule.match_value.lower() in target_value.lower()
            elif rule.match_type == "exact":
                matched = rule.match_value.lower() == target_value.lower()
            elif rule.match_type == "regex":
                try:
                    matched = bool(re.search(rule.match_value, target_value, re.IGNORECASE))
                except re.error:
                    continue

            if matched:
                txn.category_id = rule.category_id
                break

File: frontend/src/lib/api/rules.ts
import api from "./client";

export interface Rule {
  id: string;
  household_id: string;
  priority: number;
  match_field: string;
  match_type: string;
  match_value: string;
  category_id: string;
  source: string;
  enabled: boolean;
  created_at: string;
}

export interface RuleCreate {
  match_field: string;
  match_type: string;
  match_value: string;
  category_id: string;
  priority?: number;
}

export const rulesApi = {
  list: () => api.get<Rule[]>("/rules").then((r) => r.data),
  create: (data: RuleCreate) => api.post<Rule>("/rules", data).then((r) => r.data),
  update: (id: string, data: Partial<RuleCreate & { enabled: boolean }>) =>
    api.put<Rule>(`/rules/${id}`, data).then((r) => r.data),
  delete: (id: string) => api.delete(`/rules/${id}`),
};
